Add GooglePlayReleaseV4 task for uploading apks and aabs as a part of the same release

DaniilShmelev opened the pull request:
**Task name**: GooglePlayReleaseV4

**Description**: Add task for uploading apks and aabs as a part of the same release; updated UI and refactored code

# Changelog

## Common changes
- Merged code from `GooglePlayReleaseV3` and `GooglePlayReleaseBundleV3`. This V4 task should replace both V3 tasks in the future.
- Added functionality missing from `GooglePlayReleaseBundleV3`: update only store listing.
- Added functionality missing from `GooglePlayReleaseV3`: send changes to review, release name.
- Changed inputs order to better represent which inputs are general and which are additional.
- Make `applicationId` input required since we can't parse aab files easily and decided to refrain from parsing apks too.
- Removed an option to read apks without uploading them due to removing apk parsing.
- Switched to a new UI. Added the `action` input with values: `Upload single apk`; `Upload single bundle` (default); `Update only store listing`; `Upload multiple apks/aabs`. Depending on it's value inputs for uploading aab/apk files will appear and disappear. If user specifies inputs that will not be used in this action, the task will show warnings.
- Uploading mapping file is now only supported if single apk/aab is uploaded. We may add an option to upload multiple mapping files in the future.
- Simplified obb file picking up since there is no distinction between main and additional apks now. We may add an option to pick up obbs for particular apks in the future.
- Removed all hand-written google play types from `googleutil.ts`. Switched to types imported from `googleapis` library.
- Bumped `googleapis` version to simplify using `changesNotSentForReview` option in code.
- Removed `google-auth` library from dependencies since it's included in `googleapis`.
- Fixed and added jsdoc comments.
- Removed invalid `npm` demand.

## `GooglePlay.ts`
- Extracted most functions into helper files: `metadataHelper.ts`, `inputsHelper.ts`, `fileHelper.ts`. This is done because `GooglePlay.ts` was already ~900 lines long in V3.
- Split input gathering into sections for readability.
- Added support for `action` input. Introduced `Action` type. Added validation.
- Removed apk parsing.
- Simplified and refactored parts of the code.

## `googleutil.ts`
- Removed all redundant hand-written types; replaced with types from `googleapis` module
- Added functions for uploading aab, apk, oob and mapping files. Added common upload progress logging function.

**Documentation changes required:** Yes

**Added unit tests:** No (will need to implement later)

**Checklist**:
- [x] Task version was bumped - please check [instruction](https://github.com/microsoft/google-play-vsts-extension/tree/master/docs/taskversionbumping.md) how to do it
- [x] Checked that applied changes work as expected

alexander-smolyakov commented on Tasks/GooglePlayReleaseV4/GooglePlay.ts:
```diff
@@ -0,0 +1,489 @@
+
+        const actionString: string = tl.getInput('action', false);
+        if (
+            actionString !== 'MultiApkAab'
+            && actionString !== 'SingleBundle'
+            && actionString !== 'SingleApk'
+            && actionString !== 'OnlyStoreListing'
```
I believe it's better to create a set of const with all types for validating action types instead of hardcoding these checks. Also, could it make to export this validation to a separate function?

DaniilShmelev replied:
Fixed in https://github.com/microsoft/google-play-vsts-extension/pull/292/commits/8c6474660f36a9188d904a26f99fe6053d20d5bc

alexander-smolyakov commented on Tasks/GooglePlayReleaseV4/GooglePlay.ts:
```diff
@@ -0,0 +1,489 @@
+            && actionString !== 'SingleApk'
+            && actionString !== 'OnlyStoreListing'
+        ) {
+            throw new Error(`Action input value is invalid: ${actionString}`);
```
```suggestion
            throw new Error(`Action input value is invalid: ${actionString}. Please recheck pipeline task configuration.`);
```

DaniilShmelev replied:
Fixed in https://github.com/microsoft/google-play-vsts-extension/pull/292/commits/70f4a9171c4c3fefe5ab061468e4391f00fb06b2

alexander-smolyakov commented on Tasks/GooglePlayReleaseV4/GooglePlay.ts:
```diff
@@ -0,0 +1,489 @@
+        // Warn about unused inputs
+
+        switch (action) {
+            case 'MultiApkAab': warnIfUnusedInputsSet('bundleFile', 'apkFile', 'shouldUploadMappingFile', 'mappingFilePath'); break;
+            case 'SingleBundle': warnIfUnusedInputsSet('apkFile', 'bundleFiles', 'apkFiles'); break;
+            case 'SingleApk': warnIfUnusedInputsSet('bundleFile', 'bundleFiles', 'apkFiles'); break;
+            case 'OnlyStoreListing': warnIfUnusedInputsSet('bundleFile', 'apkFile', 'bundleFiles', 'apkFiles', 'track'); break;
```
Could it make sense to convert actions into objects? It allows us to encapsulate all information about each action in one place and it will make code a little bit easier to read. Also, could it make sense to combine all input arguments into a single object?

alexander-smolyakov commented on Tasks/GooglePlayReleaseV4/GooglePlay.ts:
```diff
@@ -0,0 +1,489 @@
+        tl.debug(`Checking if unused input ${input} is set...`);
+        const inputValue: string | undefined = tl.getInput(input);
+        if (inputValue !== undefined && inputValue.length !== 0) {
+            tl.warning(tl.loc('SetUnusedInput', input));
+        }
+    }
+}
+
+/**
+ * Get obb file. Returns any file with .obb extension if present in parent directory else returns
+ * from apk directory with pattern: main.<versionCode>.<packageName>.obb
+ * @param apkPath apk file path
+ * @param packageName package name of the apk
+ * @param versionCode version code of the apk
+ * @returns ObbPathFile of the obb file is present else null
+ */
+function getObbFile(apkPath: string, packageName: string, versionCode: number): string | null {
+    const currentDirectory: string = path.dirname(apkPath);
+    const parentDirectory: string = path.dirname(currentDirectory);
+
+    const fileNamesInParentDirectory: string[] = fs.readdirSync(parentDirectory);
+    const obbPathFileInParent: string | undefined = fileNamesInParentDirectory.find(file => path.extname(file) === '.obb');
+
+    if (obbPathFileInParent) {
+        tl.debug(`Found Obb file for upload in parent directory: ${obbPathFileInParent}`);
+        return path.join(parentDirectory, obbPathFileInParent);
+    }
+
+    const fileNamesInApkDirectory: string[] = fs.readdirSync(currentDirectory);
+    const expectedMainObbFile: string = `main.${versionCode}.${packageName}.obb`;
+    const obbPathFileInCurrent: string | undefined = fileNamesInApkDirectory.find(file => file.toString() === expectedMainObbFile);
+
+    if (obbPathFileInCurrent) {
+        tl.debug(`Found Obb file for upload in current directory: ${obbPathFileInCurrent}`);
+        return path.join(currentDirectory, obbPathFileInCurrent);
+    } else {
+        tl.debug(`No Obb found for ${apkPath}, skipping upload`);
+        return null;
+    }
+}
```
Could you please extract all secondary functions into separate modules? Let's try to keep only the main function in the main exec file.

DaniilShmelev replied:
Done in https://github.com/microsoft/google-play-vsts-extension/pull/292/commits/33eef3925da4f5a3ce386de1964beee789d364c1
Need to decide what to do with `prepareTrackUpdate` function

alexander-smolyakov commented on Tasks/GooglePlayReleaseV4/GooglePlay.ts:
```diff
@@ -0,0 +1,489 @@
+ *                            { track: string, versionCodes: [integer], userFraction: double }
+ */
+async function updateTrack(
+    edits: pub3.Resource$Edits,
+    packageName: string,
+    track: string,
+    versionCodes: number[],
+    versionCodeFilterType: string,
+    versionCodeFilter: string | number[],
+    userFraction: number,
+    updatePriority: number,
+    releaseNotes?: pub3.Schema$LocalizedText[],
+    releaseName?: string
```
Could it make sense to combine all these inputs into a single object?

DaniilShmelev replied:
Take a look at https://github.com/microsoft/google-play-vsts-extension/pull/292/commits/e138763b71337ce4ef23d552d164b668d5726e81
I was hesitant to do that since I don't like the look of the resulting typescript code. I could leave it or revert it. What do you think is the better option?

DaniilShmelev replied:
Reworked in https://github.com/microsoft/google-play-vsts-extension/pull/292/commits/79b751997799937003df25ca7612984ef23f11db

alexander-smolyakov commented on Tasks/GooglePlayReleaseV4/GooglePlay.ts:
```diff
@@ -0,0 +1,489 @@
+        let requireTrackUpdate = false;
+        const versionCodes: number[] = [];
+
+        if (action === 'OnlyStoreListing') {
```
Just an idea for future improvements: Considering that now the logic of the pipeline task is dependent on a selected action. Could it make sense to encapsulate all needed logic inside action itself? We even could organize a factory for actions. The benefits of this approach are following:
1) We will decrease the amount of if-else statements
2) We could modify action and worry about impacting other actions
3) It will be easier to validate tasks inputs

alexander-smolyakov commented on Tasks/GooglePlayReleaseV4/googleutil.ts:
```diff
@@ -0,0 +1,274 @@
+ * @returns track - A promise that will return result from updating a track
+ *                            { track: string, versionCodes: [integer], userFraction: double }
+ */
+export async function updateTrack(
```
As I can there is already an `async function updateTrack` function in the  `GooglePlay.ts` file
https://github.com/microsoft/google-play-vsts-extension/blob/4522cf9e8f65ec59795cfa1bb4384803dc0d3fe1/Tasks/GooglePlayReleaseV4/GooglePlay.ts#L311-L322

Could you please change the name of the function? Also, it will be great to describe the difference between functions.

DaniilShmelev replied:
Renamed in https://github.com/microsoft/google-play-vsts-extension/pull/292/commits/08ccf8f24c16b9238061d58d4530ae88777ec6ed. Not sure if there is a good name for it.
Also going to describe the difference between these functions later. Maybe a better name will come up in the process.

DaniilShmelev replied:
Reworked in https://github.com/microsoft/google-play-vsts-extension/pull/292/commits/79b751997799937003df25ca7612984ef23f11db

alexander-smolyakov reviewed (commented):
@DaniilShmelev looks very decent. Thanks for doing this and could you take a look at the comments?

Also two points:
1) Task structure. Could it make sense to move all modules to the `module` folder? Also, I believe that the main exec file should be named as `main.ts` or at least as `GooglePlayRelease.ts`.

2) Unit tests. I don't see any unit tests for the new task at the moment. Could we adopt unit tests from the V3 version of the task? I believe we need some test coverage before we will merge it.

DaniilShmelev commented:
@alexander-smolyakov Thanks! I've renamed `GooglePlay.ts` to `main.ts` and moved `googleutil.ts` and `metadataHelper.ts` to `modules` subfolder. Also migrated unit tests from GooglePlayReleaseV3.

DaniilShmelev commented on Tasks/GooglePlayReleaseV4/GooglePlay.ts:
```diff
@@ -0,0 +1,489 @@
+            let serviceEndpoint: tl.EndpointAuthorization = tl.getEndpointAuthorization(tl.getInput('serviceEndpoint', true), false);
+            key.client_email = serviceEndpoint.parameters['username'];
+            key.private_key = serviceEndpoint.parameters['password'].replace(/\\n/g, '\n');
+        }
```
Done in https://github.com/microsoft/google-play-vsts-extension/pull/292/commits/33eef3925da4f5a3ce386de1964beee789d364c1

DaniilShmelev commented:
Resolving the rest of the review points would require a major refactoring of the task, so we've decided to postpone it. Going to proceed with the current structure.

alexander-smolyakov commented on Tasks/GooglePlayReleaseV4/modules/inputsHelper.ts:
```diff
@@ -0,0 +1,179 @@
+    return key;
+}
+
+const actions = ['OnlyStoreListing', 'SingleBundle', 'SingleApk', 'MultiApkAab'] as const;
+export type Action = (typeof actions)[number];
+
+function isOfTypeAction(userInput: string): userInput is Action {
+  return (actions as readonly string[]).includes(userInput);
+}
```
Could you please add some comments regarding checking of action type?

DaniilShmelev replied:
https://github.com/microsoft/google-play-vsts-extension/pull/292/commits/e5f75f23ba4f707d16d4917cbd10fa4077978579

alexander-smolyakov reviewed (approved):
LGTM and please take a look at the comments.

DaniilShmelev commented on Tasks/GooglePlayReleaseV4/modules/fileHelper.ts:
```diff
@@ -0,0 +1,77 @@
+    } else {
+        tl.debug(`No Obb found for ${apkPath}, skipping upload`);
+        return null;
+    }
```
Resolved in https://github.com/microsoft/google-play-vsts-extension/pull/292/commits/1faf25c3890812187f4a23524560292c1550518c

DaniilShmelev commented on Tasks/GooglePlayReleaseV4/main.ts:
```diff
@@ -0,0 +1,308 @@
+            metadataRootPath = tl.getPathInput('metadataRootPath', true, true);
+        } else {
+            changelogFile = tl.getInput('changelogFile', false);
+            languageCode = tl.getInput('languageCode', false) || 'en-US';
```
Thanks for the points! Fixed in https://github.com/microsoft/google-play-vsts-extension/pull/292/commits/76bbe6c0497d1ace4e1dba6467a59f4bdb432e83
